Superchain upgrades (#1533)
ethereum-optimism/docs: pages/stack/interop/superchain-upgrades.mdx

---
title: Superchain upgrades
lang: en-US
description: Learn the superchain upgrades.
content_type: guide
topic: superchain-upgrades 
personas: 
  - node-operators
  - chain-operators
categories:
  - superchain
  - interop
  - chain operators
  - node operators
  - configuration
---

import { Callout, Tabs, Steps } from 'nextra/components'
import { InteropCallout } from '@/components/WipCallout'

# Superchain upgrades

The Superchain architecture is designed to implement coordinated, consensus-breaking changes—commonly known as hardforks—across all chains in the Superchain. 
This document provides a high-level overview of the key components and processes involved in executing these upgrades.

## Superchain target

In the context of Superchain upgrades, the **[Superchain target](https://specs.optimism.io/protocol/superchain-upgrades.html#superchain-target)** refers to a set of upgrade parameters that multiple chains in the Superchain should adhere to, enabling synchronized network upgrades. 
By aligning with a common Superchain Target, chains can implement consensus-breaking changes simultaneously, maintaining compatibility and coherence across the Superchain ecosystem. 

> This approach streamlines the upgrade process and reduces fragmentation within the superchain.

Reviewer: nice

## ProtocolVersion L1 smart contract

The **[ProtocolVersion](https://specs.optimism.io/protocol/superchain-upgrades.html#protocolversions-l1-contract)** L1 smart contract serves as an on-chain registry that records the protocol versions supported by various chains within the Superchain. 
It provides a standardized mechanism for chains to signal their current protocol version, facilitating coordination and compatibility checks among different chains. 
This contract is instrumental in managing and tracking the progression of protocol versions across the Superchain.

## Superchain version signaling

> 

Reviewer: Please add information about the `rollup.halt` flags on the CL and EL nodes:

- https://docs.optimism.io/operators/node-operators/configuration/consensus-config#rolluphalt
- https://docs.optimism.io/operators/node-operators/configuration/execution-config#rolluphalt

When these are configured, if the node see's a versioning signal onchain before the node software has been updated for a hardfork, the node will halt to prevent chain divergence

Reviewer: nvm! I see you got to it later, maybe it's worth an internal link to the section below

**[Superchain version signaling](https://specs.optimism.io/protocol/superchain-upgrades.html#superchain-version-signaling)** is a mechanism that allows nodes and clients to be aware of and prepare for upcoming protocol changes.
This signaling is achieved through the **ProtocolVersion smart contract**, allowing chains to indicate their readiness for upcoming upgrades and ensuring that all participating chains are synchronized. 
Through this signaling:
*   Nodes can anticipate and implement necessary adjustments ahead of time.
*   Clients are informed about the protocol version they should be compatible with, ensuring seamless interactions across the Superchain.

## Activation rules

**[Activation rules](https://specs.optimism.io/protocol/superchain-upgrades.html#activation-rules)** define the conditions under which a network upgrade or hardfork becomes active within the Superchain. 

> These rules can be based on specific block numbers or timestamps, dictating when the new protocol changes take effect. 

Reviewer: ```suggestion
These rules are based on specific timestamps, dictating when the new protocol changes take effect. 
```

> *   **Block Height Activation:** The upgrade activates at a predetermined block number.

Reviewer: Please denote that this is deprecated

Author: @sbvegan what are you referring to here?

Reviewer: This activation rule is no longer used, so we should either remove it or clearly denote that this method for activating hardforks is deprecated

Author: Yeah, removed it already.

*   **Time-Based Activation:** The upgrade becomes effective at a specific timestamp.

> *   **Conditional Activation:** Activation contingent upon certain network conditions or stakeholder approvals.

Reviewer: ```suggestion
```

Is this something documented in the specs? If not, I recommend removing it

## `rollup.halt` Flags on Node Binaries

The [`rollup.halt`](/operators/node-operators/configuration/execution-config#rolluphalt) flag is an opt-in configuration option available in both the execution and consensus layer node binaries. 
Its primary function is to temporarily halt node activities, during the transition phase of an upgrade, if or when it encounters an incompatible or unsupported protocol version requirement.

> *   Execution layer configuration: In the execution layer *(op-geth)*, the [`--rollup.halt`](/operators/node-operators/configuration/execution-config#rolluphalt)` flag can be set to specify the level of incompatibility (major, minor, patch, none) that will trigger a halt. This ensures that nodes do not process transactions under unsupported protocol versions.

Reviewer: ```suggestion
*   Execution layer configuration: In the execution layer *(op-geth)*, the [`--rollup.halt`](/operators/node-operators/configuration/execution-config#rolluphalt) flag can be set to specify the level of incompatibility (major, minor, patch, none) that will trigger a halt. This ensures that nodes do not process transactions under unsupported protocol versions.
```

*   Consensus layer configuration: Similarly, in the consensus layer *(op-node)*, the [`--rollup.halt`](/operators/node-operators/configuration/consensus-config#rolluphalt) flag serves the same purpose, allowing nodes to halt when encountering incompatible protocol versions.

<Callout type="warning">

> Improper use of the `rollup.halt` flag can lead to node downtime or desynchronization. 

Reviewer: ```suggestion
Failing to upgrade your node with new hardfork rules will trigger the use of the `rollup.halt` flag on the Superchain signaling and your node will halt.
```
